src/encode.rs

```diff
@@ -111,29 +116,115 @@ impl<W: io::Write> BitWriter<W> {
     }
 }
 
-/*
-/// Encode an untyped Simplicity program to bits.
+/// Encode a Simplicity program to bits, without witness data.
 ///
 /// Returns the number of written bits.
-pub fn encode_program_no_witness<'a, W: io::Write, Witness, App: Application, I>(
-    program: I,
+pub fn encode_program_no_witness<W: io::Write, Witness, App: Application>(
```

Comment:
In 0d6eb34178b483026316d1af8085f3bf90095230:

I think it'd be simpler and more efficient to inline `compute_maximal_sharing`. Then you'd only need to iterate over the program once: for each node, if it exists in `node_to_index` map then you can skip over it; otherwise serialize it in the next available position in the vector. (You are guaranteed that its children will already exist in `node_to_index` so they can be referenced.)

In bbad34e554b0950ea05d6c69a8567bc7447822b2 I think there's a similar argument for inlnining `check_maximal_sharing` into the decoding logic, though it's less strong.

Reply:
We must iterate at least two times, because the length of the program is unknown until sharing is complete. This length is the first part of the serialisation, so we must write it before writing any nodes. I do not know of any way to defer part of the serialisation without saving everything else in an inefficient vector. We have the same problem when encoding the witness, where we also iterate twice.

Reply:
Also, the `Node` references that are returned by the post-order iterator are unique. Cloned `Rc`s are already filtered out by the iterator.

Reply:
Ok, this sounds reasonable.